analytic_engine/src/compaction/picker.rs

```diff
@@ -241,51 +262,87 @@ impl LevelPicker for SizeTieredPicker {
             return None;
         }
 
-        let all_segments: BTreeSet<_> = files_by_segment.keys().collect();
         let opts = ctx.size_tiered_opts();
-
         // Iterate the segment in reverse order, so newest segment is examined first.
-        for (idx, segment_key) in all_segments.iter().rev().enumerate() {
-            // segment_key should always exist.
-            if let Some(segment) = files_by_segment.get(segment_key) {
-                let buckets = Self::get_buckets(
-                    segment.to_vec(),
-                    opts.bucket_high,
-                    opts.bucket_low,
-                    opts.min_sstable_size.as_byte() as f32,
-                );
-
-                let files = Self::most_interesting_bucket(
-                    buckets,
-                    opts.min_threshold,
-                    opts.max_threshold,
-                    opts.max_input_sstable_size.as_byte(),
-                );
-
-                if files.is_some() {
-                    info!(
-                        "Compact segment, idx: {}, size:{}, segment_key:{:?}, files:{:?}",
-                        idx,
-                        segment.len(),
-                        segment_key,
-                        segment
-                    );
-                    return files;
-                }
-                debug!(
-                    "No compaction necessary for segment, size:{}, segment_key:{:?}, idx:{}",
-                    segment.len(),
-                    segment_key,
-                    idx
-                );
+        for (idx, (segment_key, segment)) in files_by_segment.iter().rev().enumerate() {
+            let files = self.pick_ssts(segment.to_vec(), &opts);
+            if files.is_some() {
+                info!("Compact segment, idx:{idx}, segment_key:{segment_key:?}, files:{segment:?}");
+                return files;
             }
+            debug!("No compaction necessary for segment, idx:{idx}, segment_key:{segment_key:?}");
         }
 
         None
     }
 }
 
 impl SizeTieredPicker {
```

Comment:
It's not STCS now, I think maybe we should define a trait having the `pick_ssts` method and place it into TWCS?

Reply:
Although it's not strict size tiered, but I would like to keep the structure as it's now, current implementation is relative simple, we may need to group sst by size when there are no overlapping. See comment at L182.

Reply:
Moved here.